http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala

```diff
@@ -306,411 +247,267 @@ object RuntimeServiceDbQueries {
-      case Some(Right(cloudProvider))           => List(s"C.`cloudProvider` = '${cloudProvider.asString}'")
-      case None                                 => List.empty
-    }
-
-    val clusterFilters =
-      (runtimeQueryFilteredByCreator ++ runtimeQueryFilteredByDeletion ++ runtimeQueryFilteredByWorkspace ++ runtimeQueryFilteredByCloud)
-        .mkString(" AND ")
+    val runtimesFiltered = labelMap
+      .foldLeft(runtimesFilteredSimple) {
+        case (someRuntimeQuery: Query[ClusterTable, ClusterRecord, Seq], (key: String, value: String)) =>
+          for {
+            (runtime, _) <- someRuntimeQuery join labelQuery on ((r, l) =>
+              l.resourceId === r.id &&
+                l.resourceType === LabelResourceType.runtime &&
+                l.key === key &&
+                l.value === value
+            )
+          } yield runtime: ClusterTable
+      }
 
-    val clusterFiltersFinal = if (clusterFilters.isEmpty) "" else s"where ${clusterFilters}"
+    // Assemble response
+    val runtimesJoined = runtimesFiltered
+      .join(runtimeConfigs)
+      .on((runtime, runtimeConfig) => runtime.runtimeConfigId === runtimeConfig.id)
+      .joinLeft(patchQuery)
+      .on((runtimeWithConfig, runtimePatch) => runtimeWithConfig._1.id === runtimePatch.clusterId)
+      .joinLeft(labelQuery)
+      .on((runtimeWithConfigWithPatch, l) =>
+        runtimeWithConfigWithPatch._1._1.id === l.resourceId &&
+          l.resourceType === LabelResourceType.runtime
+      )
+      .map { case (((runtime, runtimeConfigRecord), runtimePatch), label) =>
+        val labelPair = label.map(l => (l.key, l.value))
+        val patchInProgress = runtimePatch.map(_.inProgress).getOrElse(false)
+        (
+          runtime.id,
+          runtime.cloudContextDb,
+          runtime.cloudProvider,
+          runtime.creator,
+          runtime.createdDate,
+          runtime.destroyedDate,
+          runtime.dateAccessed,
+          runtime.hostIp,
+          runtime.internalId,
+          patchInProgress,
+          runtimeConfigRecord,
+          runtime.runtimeName,
+          runtime.status,
+          runtime.workspaceId,
+          labelPair
+        )
+      }
 
-    val labelMapFilters =
-      if (labelMap.isEmpty)
-        ""
-      else {
-        val query = labelMap
-          .map { case (k, v) =>
-            s"(LABEL.key = '${k}' and LABEL.value = '${v}')"
+    runtimesJoined.result
+      .map { records =>
+        type ListRuntimesRecord = (
+          Long,
+          CloudContextDb,
+          CloudProvider,
+          WorkbenchEmail,
+          Instant,
+          Instant,
+          Instant,
+          Option[IP],
+          String,
+          Boolean,
+          RuntimeConfigRecord,
+          RuntimeName,
+          RuntimeStatus,
+          Option[WorkspaceId],
+          Option[(String, String)]
+        )
+        records
+          .groupBy(_._1)
```

Comment:
`id`

Reply:
is there a way to refer to these things by "name" vs position?

Reply:
correction ... this is a scala `groupBy` vs a slick `groupBy` - so we should only be looking at a smaller list.

ask is to verify that the cardinality of the result set is small (filtered to a specific user)